crates/driver/src/domain/competition/solution/scoring.rs

```diff
@@ -209,6 +181,102 @@ impl Trade {
         })
     }
 
+    /// Protocol fee as a cut of surplus, denominated in SURPLUS token
+    ///
+    /// Protocol fee calculation logic depends if the protocol fee
+    /// is already applied to the trade or not. Since scoring module works with
+    /// trades that already have the protocol fee applied, we need to calculate
+    /// the protocol fee as an observation of already applied protocol fee.
+    ///
+    /// The protocol fee before being applied is:
+    ///    fee = surplus_before_fee * factor
+    /// The protocol fee after being applied is:
+    ///    fee = surplus_after_fee * factor'
+    /// Also:
+    ///    surplus_after_fee = surplus_before_fee - fee
+    /// So:
+    ///    factor' = fee / surplus_after_fee = fee / (surplus_before_fee - fee)
+    /// = fee / ((fee / factor) - fee) = factor / (1 - factor)
+    ///
+    /// Finally:
+    ///     fee = surplus_after_fee * factor / (1 - factor)
+
+    fn surplus_fee(&self, factor: f64) -> Result<eth::Asset, Error> {
+        let surplus = self.surplus().ok_or(Error::Surplus(self.sell, self.buy))?;
+        let fee = surplus
+            .amount
+            .apply_factor(factor / (1.0 - factor))
+            .ok_or(Math::Overflow)?;
+
+        Ok(eth::Asset {
+            token: surplus.token,
+            amount: fee,
+        })
+    }
+
+    /// Protocol fee as a cut of the trade volume, denominated in SURPLUS token
+    ///
+    /// Protocol fee calculation logic depends if the protocol fee
+    /// is already applied to the trade or not. Since scoring module works with
+    /// trades that already have the protocol fee applied, we need to calculate
+    /// the protocol fee as an observation of already applied protocol fee.
```

Comment:
```suggestion
    /// Volume fee is specified as a `factor` from raw volume (before fee). Since this module works with
    /// trades that already have the protocol fee applied, we need to calculate
    /// the protocol fee as an observation of a the eventually traded amount using a different factor `factor'` .
```

Reply:
copied